File: Day5_Next Permutation/next_permutation.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Scanner;

public class next_permutation {
    public static ArrayList<Integer> get_permutation(ArrayList<Integer> arr) {
        int break_point = -1;
        for (int i = arr.size()-2; i >= 0 ; i--) { // step 1
            if (arr.get(i) < arr.get(i+1)) {
                break_point = i;
                break;
            }
        }
        if(break_point==-1) {
            Collections.reverse(arr);
            return arr;
        }

        for (int i = arr.size()-1; i > break_point ; i--) { // step 2
            if (arr.get(i) > arr.get(break_point)) {
                Collections.swap(arr, break_point, i);
                break;
            }
        }
        List<Integer> sublist = arr.subList(break_point + 1, arr.size()); // Step 3
        Collections.reverse(sublist);
        return arr;
    }
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.println("Enter the size and space separated array: ");
        int n = sc.nextInt();
        ArrayList<Integer> arr = new ArrayList<Integer>();
        for (int i = 0; i < n; i++) {
            arr.add(sc.nextInt());
        }
        ArrayList<Integer> result = get_permutation(arr);
        for (int num : result) {
            System.out.print(num + " ");
        }
        sc.close();
    }
}

File: Day5_Next Permutation/next_permutation.py
# https://www.geeksforgeeks.org/batch/gfg-160-problems/track/arrays-gfg-160/problem/next-permutation5226

def nextPermutation(arr):
    # code here
    break_point = -1 # 1.check where the decremental ordering is breaking
    for i in range(len(arr) - 2, -1, -1):
        if arr[i] < arr[i + 1]:
            break_point = i
            break
    if break_point == -1: # it is the last permutation
        arr.reverse()
        return arr

    for i in range(len(arr) - 1, break_point, -1): # 2.swap break point with the next greater number
        if arr[i] > arr[break_point]:
            arr[break_point], arr[i] = arr[i], arr[break_point]
            break
    arr[break_point + 1 :] = reversed(arr[break_point + 1 :]) # 3.reverse the element from break point to end
    return arr

user_input = input("Enter space-separated integers: ")
arr = list(map(int, user_input.split()))
print(nextPermutation(arr=arr))
